### ISSUE TITLE
Switch Python rules to use a proper starlark-accessible provider

### ISSUE THREAD

**Issue by brandjon** [bazelbuild/bazel#7010]
The `py` legacy-style string-named provider should be replaced by a `PyInfo` provider type, whose fields are enumerated and typed. This should also subsume the native `PythonImportsProvider` type.

This should more or less implement "Python sandwich" -- having Starlark-defined-rules depend on the native Python rules and vice versa.

- [x] Add helper to abstract over which provider type is used and default value logic
- [x] Implement the new provider type, make Python rules consume and return it
- [x] Add incompatible flag for making Python rules disallow and not return the old provider
- [x] Make migration-ready
- [x] Migrate Bazel's own tests
- [x] Migrate downstream and flip the flag

**Comment by brandjon** [comment]
Note that some form of Starlark sandwich is already possible. For instance, I can write a py_library replacement that exposes its srcs in a `transitive_sources` field. We need to document an exact API.

To accelerate/unblock work on migrating away from the legacy struct providers, we might decouple the change in the provider type from the change in the provider's fields. I.e., we could have one incompatible change flag to deprecate `ctx.attr.my_dep.py` with `ctx.attr.my_dep[PyInfo]`, and another incompatible change flag that affects the names or types of a bunch of fields on PyInfo.

**Comment by brandjon** [comment]
Related issues: #7054 and #7121.

**Comment by brandjon** [comment]
Will work on this next. For the purposes of this tracking bug, the contents of the provider are out-of-scope; we only care about accessing it as a modern provider instead of a legacy one.

**Comment by brandjon** [comment]
Change of plan: The incompatible flag will actually *disallow* passing the legacy provider to native Python rules, not just make it ignored. Failing fast makes it easier to find rules that need migrating, and avoids subtle bugs due to falling back on default logic.

**Comment by brandjon** [comment]
At this point the feature is fully implemented (to be in 0.23), and this issue is now blocked on migration and flipping the incompatible change flag (#7298), then cleanup of the legacy code paths.

**Comment by brandjon** [comment]
Cleanup of legacy code tracked by #7741.
